### PR TITLE
Make evp_test memory compares error aware.

### PR DESCRIPTION
**Pull request by paulidale**
Add a helper routine so that evp_test can compare memory without producing spurious output when checking for error conditions.

- [x] tests are added or updated

Refer #6857
An alternative to #6862

### CODE REVIEW

File: test/evp_test.c [hunk 1]
```diff
@@ -74,6 +74,27 @@ static int find_key(EVP_PKEY **ppk, const char *name, KEY_LIST *lst);
...
+static int memory_err_compare(EVP_TEST *t, const char *err,
+                              const void *expected, size_t expected_len,
+                              const void *got, size_t got_len)
+{
+    int r;
+
+    if (t->expected_err != NULL && strcmp(t->expected_err, err) == 0)
+        r = !TEST_mem_ne(expected, expected_len, got, got_len);
+    else
+        r = TEST_mem_eq(expected, expected_len, got, got_len);
```
Comment (mspncp) [inline test/evp_test.c]: All error messages will be logged with line number 90 or 92.

**Comment by mspncp** [comment]
Note: I renamed `memory_err_compare` to `test_mem_err` and changed also the parameter names of the latter in order to get more analogy to the names used in `testutil/testutils.h` and `testutil/tests.c`. So the test macro now is `TEST_mem_err` and it's implementation is `test_mem_err`.

**Comment by mspncp** [comment]
> So the test macro now is `TEST_mem_err` and it's implementation is `test_mem_err`.

An alternative would be to name them `TEST_evp_mem_err` and `test_evp_mem_err`, respectively.

File: test/evp_test.c (hunk 1 above)
Comment (paulidale) [reply test/evp_test.c]: I wasn't too concerned about this.  The output that `evp_test` already produces uniquely determines the failure line -- search for the error or look at the kind of algorithm being run.  Either describes it.

However, I'm dubious about tests using the internals of testutil.  The macros are the defined interface, the underlying functions aren't intended to be called directly.

Does it make sense to add TEST_mem_eq_err and friends to the framework?  I.e. tests which are expecting an error and produce output when there isn't one?  I'm not convinced they'd be sufficiently useful.

Comment (mspncp) [reply test/evp_test.c]: I checked it again, you are right. The improvement of the output gained by my patch is very small. So forget about my suggestion.

**Comment by paulidale** [comment]
Merged, thanks.

**Pull request closed without merging**